superbench/benchmarks/micro_benchmarks/ib_validation_performance/ib_validation_performance.cc

```diff
@@ -0,0 +1,450 @@
+}
+
+// Load and parse config file to vector
+vector<vector<std::pair<int, int>>> load_config(string filename = "config.txt") {
+    // read contents from file
+    vector<std::string> config;
+    ifstream in(filename);
+    string line;
+
+    if (in) {
+        while (getline(in, line)) {
+            config.push_back(line);
+        }
+    } else {
+        std::cout << "Error: failed to open config file." << std::endl;
+        MPI_Finalize();
+    }
+
+    if (world_rank == 0) {
+        std::cout << "config: " << std::endl;
+        for (const std::string &line : config) {
+            std::cout << line << std::endl;
+        }
+    }
+
+    // parse the string contents to vector
+    vector<vector<std::pair<int, int>>> run_in_total;
+    try {
+        for (const auto &single_line : config) {
+            // parse each line like "1,2;2,3;3,4" to vector<std::pair<int, int>>
+            vector<string> run_in_parallel;
+            vector<std::pair<int, int>> run_pairs_in_parallel;
+            // split line to pair by ";"
+            split_string(single_line, run_in_parallel, ";");
+            for (const auto &pair : run_in_parallel) {
+                // split pair by ","
+                int quote = pair.find(',');
+                int first = stoi(pair.substr(0, quote));
+                int second = stoi(pair.substr(quote + 1));
+                run_pairs_in_parallel.emplace_back(first, second);
+            }
+            run_in_total.emplace_back(run_pairs_in_parallel);
+        }
+    } catch (const std::exception &e) {
+        std::cout << "Error: failed to parse the config, message: " << e.what() << std::endl;
+        MPI_Finalize();
+    }
+
+    return run_in_total;
+}
+
+// Execute shell cmd in termial and get output
+std::string exec(const char *cmd) {
+    char buffer[128];
+    std::string result = "";
+    // use pipe to execute command
+    FILE *pipe = popen(cmd, "r");
+    if (!pipe)
+        throw std::runtime_error("popen() failed!");
+    try {
+        // use buffer to get output
+        while (fgets(buffer, sizeof buffer, pipe) != NULL) {
+            result += buffer;
+        }
+    } catch (const std::exception &e) {
+        pclose(pipe);
+        throw e;
+    }
+    pclose(pipe);
+    return result;
+}
+
+// Get private ip of current machine
+std::string get_my_ip() {
+    std::string ip;
+    try {
+        string command = "hostname -I | awk '{print $1}'";
+        ip = exec(command.c_str());
+        int r = ip.find('\n');
+        ip = ip.substr(0, r);
+    } catch (const std::exception &e) {
+        std::cout << "Error: failed to get ip of " << processor_name << std::endl;
+        MPI_Finalize();
+    }
+    return ip;
+}
+
+// Use socket to get a free port
+int get_available_port() {
+    struct sockaddr_in server_addr;
+    int sockfd;
+    // open a socket
+    if ((sockfd = socket(AF_INET, SOCK_STREAM, 0)) < 0) {
+        std::cout << "Error: failed to open socket.";
+        MPI_Finalize();
```

Comment:
It's a little bit strange for `MPI_Finalize()` here. Because all operation for MPI is in the main().